File: mobile/src/domain/substitutes.ts
/**
 * Approved substitutes of a planned slot, read from its locked notes, and the rules for an
 * exercise name the lifter types. Mirrors the desktop's `domain/substitutes.py` and
 * `domain/exercise_names.py` exactly.
 */

export type ApprovedSubstitute = {
  name: string;
  /** The source's own condition for this substitute ("if unavailable/intolerant"), if any. */
  condition: string | null;
};

const PREFIX = 'Approved substitutes:';
const CONDITION = ' if ';

/** The slot's approved substitutes in source order; none when its notes list none. */
export function approvedSubstitutes(notes: string | null): ApprovedSubstitute[] {
  if (notes === null) return [];
  for (const line of notes.split(/\r?\n/)) {
    const text = line.trim();
    if (!text.startsWith(PREFIX)) continue;
    let body = text.slice(PREFIX.length).trim();
    if (body.endsWith('.')) body = body.slice(0, -1);
    const found: ApprovedSubstitute[] = [];
    for (const item of body.split(', ')) {
      const entry = item.trim();
      if (!entry) continue;
      const at = entry.indexOf(CONDITION);
      found.push(
        at === -1
          ? { name: entry, condition: null }
          : {
              name: entry.slice(0, at).trim(),
              condition: `if ${entry.slice(at + CONDITION.length).trim()}`,
            },
      );
    }
    return found;
  }
  return [];
}

export const MAX_EXERCISE_NAME_LENGTH = 80;
const WORD_BREAKS = new Set([' ', '-', '/', '(']);

/** Comparison key that keeps trivial duplicates out: case- and whitespace-insensitive. */
export function exerciseNameKey(name: string): string {
  return name.split(/\s+/).filter(Boolean).join(' ').toLowerCase();
}

export type TypedName = { ok: true; name: string } | { ok: false; error: string };

/**
 * A typed exercise name: surrounding and repeated whitespace goes; an all-lowercase entry is
 * title-cased ("triceps curl" → "Triceps Curl"); any other casing is the lifter's own.
 */
export function typedExerciseName(text: string): TypedName {
  const collapsed = text.split(/\s+/).filter(Boolean).join(' ');
  if (!collapsed) return { ok: false, error: 'Type the exercise’s name.' };
  if (collapsed.length > MAX_EXERCISE_NAME_LENGTH) {
    return { ok: false, error: `At most ${MAX_EXERCISE_NAME_LENGTH} characters.` };
  }
  if (collapsed !== collapsed.toLowerCase()) return { ok: true, name: collapsed };
  let previous = ' ';
  let name = '';
  for (const ch of collapsed) {
    name += WORD_BREAKS.has(previous) ? ch.toUpperCase() : ch;
    previous = ch;
  }
  return { ok: true, name };
}

File: backend/src/fitness_lab/domain/substitutes.py
"""Approved substitutes of a planned slot, read from its locked notes.

The locked program lists them per exercise (``substitution_matrix``); the adapter writes them
verbatim into each slot's notes as one line, ``Approved substitutes: A, B, C.``. This module
only reads that line back. A trailing condition in the source ("Lying Leg Curl if
unavailable/intolerant") is kept, split from the exercise name so the name can be an exercise
identity while the condition is still shown.
"""

from __future__ import annotations

from dataclasses import dataclass

PREFIX = "Approved substitutes:"
CONDITION = " if "


@dataclass(frozen=True, slots=True)
class ApprovedSubstitute:
    name: str
    # The source's own condition for this substitute ("if unavailable/intolerant"), if any.
    condition: str | None


def approved_substitutes(notes: str | None) -> tuple[ApprovedSubstitute, ...]:
    """The slot's approved substitutes in source order; none when the notes list none."""
    if notes is None:
        return ()
    for line in notes.splitlines():
        text = line.strip()
        if not text.startswith(PREFIX):
            continue
        body = text[len(PREFIX) :].strip().removesuffix(".")
        found: list[ApprovedSubstitute] = []
        for item in body.split(", "):
            entry = item.strip()
            if not entry:
                continue
            name, separator, condition = entry.partition(CONDITION)
            found.append(
                ApprovedSubstitute(
                    name=name.strip(),
                    condition=f"if {condition.strip()}" if separator else None,
                )
            )
        return tuple(found)
    return ()
